### DISCUSSION TITLE
How can I point leapp at custom mirrors for the target OS/version?

### DISCUSSION THREAD

**Discussion by verdverm** [oamg/leapp-repository#1189] [Q&A] [answered]
We run our systems in an restricted environment where we can only access yum mirrors we host. Trying to do a Centos7 -> Rocky8 upgrade.

Is there a place where we can set the repo locations for the upgrade process? Which actor(s) is handling this?

I noticed the following get automatically setup during `leapp preupgrade`

```
====> * target_userspace_creator
        Initializes a directory to be populated as a minimal environment to run binaries from the target system.
Rocky Linux 8 - PowerTools                       32 MB/s | 2.7 MB     00:00    
Rocky Linux 8 - ResilientStorage                 14 MB/s | 792 kB     00:00    
Rocky Linux 8 - Extras                          558 kB/s |  14 kB     00:00    
Rocky Linux 8 - AppStream                        85 MB/s |  11 MB     00:00    
Rocky Linux 8 - HighAvailability                 16 MB/s | 784 kB     00:00    
Rocky Linux 8 - BaseOS                           65 MB/s | 7.1 MB     00:00    
```

**Comment by pirat89** [comment, marked as answer]
Hi, it should be `tartget_userspace_actor` as in the output. Based on your question, I expect that you haven't defined following target repositories manually. In such a case you are most likely using an Elevate project or a different modified downstream version of leapp-repository (as this project currently supports only upgrades of RHEL systems and does not provide cross-distro conversion), so I cannot tell you how to correctly deal with your situation. I would check whether somewhere in installed packages are listed rocky8 repo files. If so, most likely you can just update definition of repositoires in such files to deal with it. But there is also possiiblity that repofiles are downloaded or created dynamically. 

Our solution for RHEL expects e.g. `leapp preupgrade --no-rhsm` and defined `/etc/leapp/files/leapp_upgrade_repositories.repo` repo file ([doc](https://access.redhat.com/articles/4977891)). But  I expect the solution is different in your packages as RHSM (which we use to obtain target RHEL repositories) is specific to RHEL systems. For the proper instructions you need to ask vendor of packages you have installed. I would expect them to have it documented somewhere if they provide this possibility. In case you installed packages from the Elevate project, I am aware there is a discussion forum where they deal with such things. I would not be surprised if someone asked there such a question in the past (I would search keywords like: offline, disconnected, custom repositories).

Update: 
* from other comments: https://wiki.almalinux.org/elevate/ELevate-offline-guide.html

marking as resolved

**Comment by verdverm** [comment]
@pirat89 Thanks for the pointers, we are indeed using Elevate from Alma

`https://repo.almalinux.org/elevate/el7/$basearch/`

I'm guessing https://almalinux.discourse.group/ might be the forum you are thinking of based on https://wiki.almalinux.org/ ?

**Comment by verdverm** [comment]
Discovered this: https://wiki.almalinux.org/elevate/ELevate-offline-guide.html

**Comment by verdverm** [comment]
@pirat89 Do you know if we can add additional repositories to the upgrade list, would this enable us to upgrade more packages from 3rd party repositories?

https://github.com/AlmaLinux/leapp-data/blob/main/files/rocky/leapp_upgrade_repositories.repo.el8

**Reply by pirat89** [reply]
at least in our upstream solution, yes you can. regarding third-party packages, it's even expected that users will do it. in case that vendors of third-party content did packaging right, most likely packages will be updated correctly during the upgrade. but in some cases it's possible that you would need to develope own custom leapp actors to adjust/customize the process. you can use either the `--enablerepo <repoid>` option (can be used multiple times, once per each repoid) when running leapp (in such a case, the repository must be already defined under /etc/yum.repos.d directory). another way is the one i described in my previous comment. you can define as many repositories inside the file as you want. only requirement is that the repoid must be unique, so no duplicates definitions exists under /etc/yum.repos.d/ during the upgrade process.

**Reply by verdverm** [reply]
can the "before" (centos-7) repos be arbitrarily ID'd?

(i.e. we use `c7-base` and `r8-baseos` for the next version)

**Reply by pirat89** [reply]
yes. so far, when you specify manually what repositories should be used during the upgrade (the option or the file mentioned earlier), you can use whatever repoids you want if they are considered to be valid by DNF. means, whatever ascii string without use of variables (e.g. this is invalid: `el$releasever-baseos`; but `el8-baseos` is valid)

**Reply by verdverm** [reply]
To clarify, the upgrade process is not looking for specific ids on the centos-7 repos, correct?

Do the new repos need to be accessible before the reboot, and they get fetched and installed during that initial phase after reboot?

Am I understanding the order of operations correctly?

1. use old repos
2. `leap upgrade` and reboot
3. use new repos

**Reply by pirat89** [reply]
no, repositories must be accessible before the reboot. as part of the preupgrade process is to calculate the RPM upgrade transaction (for which it must be used el8 DNF+RPM - downloded from target repositories), download ALL required packages and test the transaction - e.g. for file conflicts, etc). 

> to clarify, the upgrade process is not looking for specific ids on the centos-7 repos, correct?

not sure. in case of RHEL, we are checking also old rhel7 repos and installed content to know what rhel8 repos should be enabled. in case of centos, there are not so many options. And also, I think you are going to use the mentioned repo file or --enablerepo option, so it should not affect you so much. most likely only requirement it, that at least one centos 7 repo is enabled before running leapp, so `yum clean all` does not fail.

**Comment by pirat89** [comment]
Seems the discussion is resolved. Closing the topic.

**Reply by verdverm** [reply]
yup, I'll open a new issue/discussion should I need to
